torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,306 @@
+from typing import Any, Callable, List, Optional
+
+import torch
+import torch.distributed as dist
+from torch.distributed.optim import ZeroRedundancyOptimizer
+from torch.distributed.optim.zero_redundancy_optimizer import _OverlapStatus
+from torch.nn.parallel.distributed import DistributedDataParallel
+
+
+# Functional optimizers require passing a list of gradients to their `step()`
+# method, and ZeRO requires a functional optimizer to overlap with DDP
+# Passing a `None` instead of an actual gradient indicates to the optimizer
+# to not update the corresponding parameter
+_NO_PARAM_UPDATE = None
+
+
+def hook_with_zero_step(
+    hook: Callable[[Any, dist.GradBucket], torch.futures.Future],
+    ddp: DistributedDataParallel,
+    zero: ZeroRedundancyOptimizer,
+) -> Callable[[Any, dist.GradBucket], torch.futures.Future]:
+    r"""
+    Modifies the given ``hook`` to overlap the :class:`ZeroRedundancyOptimizer`
+    optimizer step with the :class:`DistributedDataParallel` backward pass,
+    where the optimizer step computation begins after the last gradient bucket
+    computation has finished.
+
+    This approach overlaps the optimizer computation and communication with the
+    backward communication. In particular, the backward computation proceeds
+    contiguously, and the optimizer computation follows, overlapping with
+    outstanding backward communication (i.e. all-reduces) and possibly other
+    optimizer communication (i.e. broadcasts).
+
+    This approach may be preferred over :meth:`hook_with_zero_step_interleaved`
+    if communication is relatively slow compared to computation.
+
+    Arguments:
+        hook (Any * dist.GradBucket -> torch.futures.Future): the hook to
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+
+    Raises:
+        ValueError: if ``zero`` was constructed with ``overlap_with_ddp=False``.
+    """
+    if not zero._overlap_with_ddp:
+        raise ValueError(
+            "ZeroRedundancyOptimizer must be constructed with "
+            "`overlap_with_ddp=True` to use this hook properly"
+        )
+
+    def hook_with_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs the equivalent of a :class:`ZeroRedundancyOptimizer`
+        :meth:`step` if ``bucket`` is the last gradient bucket.
+
+        The function performs additional computation on the iteration that
+        the :class:`DistributedDataParallel` buckets are rebuilt to collect
+        information used to implement the modified hook.
+
+        Arguments:
+            state: any state for the hook.
```

Comment:
Fixed.

Reply:
Actually, I was reading this: https://stackoverflow.com/questions/39817081/typing-any-vs-object

Even though the type annotation for `state` in `register_comm_hook()` in `distributed.py` is `object`, I wonder if that should be `Any` instead. If I am understanding correctly, if you type annotate `state` with `object`, then to appease a type checker, every operation involving `state` must work for any general `object` type. In other words, annotating with `object` is actually **restrictive** rather than all-encompassing.